### ISSUE TITLE
Change in semantics for the device plugin registration process

### ISSUE THREAD

**Issue by klueska** [kubernetes/kubernetes#112395]
### What happened?

The following refactoring of the kubelet's devicemanager was done as part of the 1.25 release: https://github.com/kubernetes/kubernetes/commit/db88676c20629354d8eeb4b7014ae7884481ffca

As part of this, a change in semantics was (unintentionally) introduced which is breaking at least one 3rd party plugin. The logic required to go back to the old semantics is fairly straight forward. However, it's not obvious that we should actually go back to the old semantics because (at least in my opinion) the new semantics are actually more correct.

Previously, the kubelet's devicemanager would do the following upon receiving a plugin registration request:
1. Register the plugin
2. Launch a go-routine to connect to the gRPC service being served by the plugin (with a timeout of 10s)
3. Return

This gave the plugin the opportunity to first register itself with the kubelet and then (within a 10s time window) start its gRPC server.

The new semantics are similar, except that no go-routine is launched to connect to the gRPC service. The connection is attempted synchronously. This means that the plugin must now start serving its gRPC server *before* registering itself with the kubelet. Otherwise the registration call will fail.

As mentioned before, this change was not necessarily intentional, but it's also not clear if it should be reverted. It brings the device plugin's custom registration process more in line with the kubelet's standard registration process and the order is actually a bit more intuitive. Why bother registering if you aren't able to start serving your API?

The first plugin to report this problem was the KubeVirt plugin and a discussion of the issue can be found here:
https://groups.google.com/g/kubevirt-dev/c/HE2lVvsLd3Y

The question now is -- should we revert the kubelet code to the old semantics to avoid breaking plugins that rely on this (somewhat awkward) ordering. Or do we leave the new semantics in place and send out an announcement of this (potentially) breaking change?

One thing to keep in mind -- if we go with the second option (i.e. keep the new semantics and force plugins to update), the updated plugins will continue to work on older versions of kubernetes since the ordering now implied in the new semantics has always been correct.

### What did you expect to happen?

Plugins relying on the old semantics would continue to work on kubernetes 1.25

### How can we reproduce it (as minimally and precisely as possible)?

Write a plugin that:
1. Registers itself with the kubelet
2. Only then starts to server the plugin API via a gRPC server

### Kubernetes version

<details>
1.25
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by fromanirh** [comment]
xref: https://github.com/kubevirt/kubevirt/pull/8451

**Comment by fromanirh** [comment]
I'm interested in the fact that the e2e tests didn't seem to catch this issue. Besides improving our coverage, which rarely hurts, I tend to believe that one of the key factors is that we use sample-device-plugin in the e2e tests, which is understandably updated in lockstep with the core k/k code. I wonder how we can improve in this area, perhaps using _also_ a known set of 3rd party device plugins as part as e2e testing?

**Comment by klueska** [comment]
The plugin used in the e2e tests uses the standard kubelet plugin registration method (i.e. the one that relies on the plugin dropping a socket file in a well known location to trigger the registration), not the custom / legacy one used by the KubeVirt (and potentially many other) plugins.

**Comment by fromanirh** [comment]
TL;DR: if the broken plugins were doing legitimate things (-> were not themselves buggy) we should fix the kube side, even if the new process really seems better.

My initial though is this is the right change done the wrong way. The new registration process should be enabled - for all the good reason @klueska listed - but we should have better docs and a transition process (e.g. deprecating the old process, supporting both for a little while... details to be figured out).

Now: the device manager is stil a [beta feature in 1.25](https://github.com/kubernetes/kubernetes/blob/v1.25.0/pkg/features/kube_features.go#L247) (!!!!!!) so we have some slack space, but still this change breaks some device plugins which, AFAIK, were not buggy in the area of the registration process (is my understanding right?). This is not ok for a beta feature AFAIK, and not great in general.

I reserve myself to review in deeper detail and come back, but IMO we should treat this like a bug and revert the change, *then* work towards the better registration process.

**Comment by mythi** [comment]
> but still this change breaks some device plugins which, AFAIK, were not buggy in the area of the registration process (is my understanding right?)

FWIW, the sequence is [documented](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation) and it says the plugin starts a gRPC service and then register itself.

**Comment by cynepco3hahue** [comment]
Agree with @fromanirh at least we should give time and proper notification to device plugin owners, so:
1. revert it for now
2. add a deprecation message under device manager, that will inform plugins owners to update their device plugins(and probably we also should have updated documentation that will explain the correct way to initialize now the device plugin)
3. revert the revert after one or two releases :)

**Comment by mythi** [comment]
1.-3. sounds overkill if it's just one plugin that was not following the doc and needs fixing.

**Comment by swatisehgal** [comment]
> As part of this, a change in semantics was (unintentionally) introduced which is breaking at least one 3rd party plugin. The logic required to go back to the old semantics is fairly straight forward. However, it's not obvious that we should actually go back to the old semantics because (at least in my opinion) the new semantics are actually more correct.

I tend to agree that the new semantics (starting gRPC server before registration) are more intuitive and correct.
> The question now is -- should we revert the kubelet code to the old semantics to avoid breaking plugins that rely on this (somewhat awkward) ordering. Or do we leave the new semantics in place and send out an announcement of this (potentially) breaking change?

Based the following points, I am inclining towards the latter (sticking to the new semantics and updating device plugins impacted by this change):

- the documentation already captures the correct ordering
- the plugin that brought this issue to light has already been updated
- a plugin once updated would continue to work on older versions of Kubernetes

Reverting the change seems like a step back when we can handle this by sending out communication to highlight the potential issue and capturing explicitly in the documentation that "a device plugin **must** start serving gRPC server before registering itself".

**Comment by fromanirh** [comment]
> So @mythi and @swatisehgal pointed out that the correct ordering was already documented. Then the basic assumption in my previous [#112395 (comment)](https://github.com/kubernetes/kubernetes/issues/112395#issuecomment-1246475947) is void. I'll have a deeper review myself ASAP, but if this is the case, then retroactively fixing the docs and sending a late announcement seems indeed the best approach, and reverting the changes turns from "overkill" to "actually wrong". I'll update ASAP.

So the[ device plugin flow](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-implementation) was documented in its current form (which is not broken by the change we're discussing) [since 2018](https://github.com/kubernetes/website/commit/169aadbbe185b6c42c8f2a4e8066548903b72d12)

Then yes, my initial reaction is overtuned. It would probably have be nice to give the device plugin author some leniency before to enforce the correct flow as we did by accident, but the fact the kube code enforces the documented, expected flow still stands. So we better handle this issue with a documentation update and/or a release note update.
